File: 16. Search in Sorted Matrix.cpp
// Topic: Search in row & column-wise sorted matrix (staircase search)
#include <bits/stdc++.h>
using namespace std;
bool searchMatrix(const vector<vector<int>>& A, int target)
{
    if (A.empty() || A[0].empty())
        return false;
    int n = A.size(), m = A[0].size();
    int i = 0, j = m - 1;
    while (i < n && j >= 0)
    {
        if (A[i][j] == target)
            return true;
        if (A[i][j] > target)
            --j;
        else
            ++i;
    }
    return false;
}
int main()
{
    vector<vector<int>> A = {{1, 4, 7, 11}, {2, 5, 8, 12}, {3, 6, 9, 16}};
    cout << boolalpha << searchMatrix(A, 8) << "\n";  // true
}

// TC: O(n+m)
// SC: O(1)

File: 16. Search in Sorted Matrix.py
"""Topic: Search in row & column-wise sorted matrix"""
def search(A,target):
    if not A or not A[0]: return False
    n,m=len(A),len(A[0])
    i,j=0,m-1
    while i<n and j>=0:
        if A[i][j]==target: return True
        if A[i][j]>target: j-=1
        else: i+=1
    return False

if __name__=='__main__':
    A=[[1,4,7,11],[2,5,8,12],[3,6,9,16]]
    print(search(A,8))

# TC: O(n+m)
# SC: O(1)
